cdap-security/src/main/java/co/cask/cdap/security/authorization/DefaultAuthorizationEnforcementService.java

```diff
@@ -0,0 +1,155 @@
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * Default implementation of {@link AuthorizationEnforcementService}.
+ */
+@Singleton
+public class DefaultAuthorizationEnforcementService extends AuthorizationEnforcementService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultAuthorizationEnforcementService.class);
+
+  private final AuthorizerInstantiator authorizerInstantiator;
+  private final boolean authorizationEnabled;
+  private final boolean cacheEnabled;
+  private final int cacheRefreshIntervalSecs;
+  private final LoadingCache<Principal, Set<Privilege>> authPolicyCache;
+  private final Lock readLock;
+  private final Lock writeLock;
+
+  @Inject
+  public DefaultAuthorizationEnforcementService(final AuthorizerInstantiator authorizerInstantiator,
+                                                CConfiguration cConf) {
+    this.authorizerInstantiator = authorizerInstantiator;
+    this.authorizationEnabled = cConf.getBoolean(Constants.Security.Authorization.ENABLED);
+    this.cacheEnabled = cConf.getBoolean(Constants.Security.Authorization.CACHE_ENABLED);
+    this.cacheRefreshIntervalSecs = cConf.getInt(Constants.Security.Authorization.CACHE_REFRESH_INTERVAL_SECS);
+    this.authPolicyCache = CacheBuilder.newBuilder()
+      .expireAfterWrite(cConf.getInt(Constants.Security.Authorization.CACHE_TTL_SECS), TimeUnit.SECONDS)
+      .build(new CacheLoader<Principal, Set<Privilege>>() {
+        @SuppressWarnings("NullableProblems")
+        @Override
+        public Set<Privilege> load(Principal key) throws Exception {
+          return authorizerInstantiator.get().listPrivileges(key);
+        }
+      });
+    ReadWriteLock lock = new ReentrantReadWriteLock();
+    this.readLock = lock.readLock();
+    this.writeLock = lock.writeLock();
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    return Scheduler.newFixedDelaySchedule(0, cacheRefreshIntervalSecs, TimeUnit.SECONDS);
+  }
+
+  @Override
+  protected void runOneIteration() throws Exception {
+    if (!authorizationEnabled || !cacheEnabled) {
+      return;
+    }
+    LOG.debug("Running authorization enforcement service iteration...");
+    // we want to block all access to the cache while the cache update is running
+    writeLock.lock();
+    try {
+      for (Principal principal : authPolicyCache.asMap().keySet()) {
+        updatePrivileges(principal);
+      }
+    } finally {
+      writeLock.unlock();
+    }
+  }
+
+  @Override
+  public void updatePrivileges(Principal principal) throws Exception {
+    if (!authorizationEnabled || !cacheEnabled) {
+      return;
+    }
+    Authorizer authorizer = authorizerInstantiator.get();
+    Set<Privilege> privileges = authorizer.listPrivileges(principal);
```

Comment:
but that will not update a principal's privileges if the entry in the cache has not expired, right? This method is meant to always update the cache entry, no matter if the entry is expired or not.